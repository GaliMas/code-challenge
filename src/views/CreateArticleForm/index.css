.createArticleForm .labelForm{
    display: block;
}

.createArticleForm input[type="text"], .createArticleForm textarea{
  width: 500px;
}
.createArticleForm textarea{
  resize: none;
}

.createArticleForm button{
  display: inline-block;
  height: 36px;
  padding: 0 2rem;
  position: relative;
}

.createArticleForm.cancel{
  float: left;
}
.createArticleForm .save{
  float:right;
}

.createArticleForm .formRow{
  margin: 10px 0;
}

.createArticleForm .tagContainer{
  border: 1px solid;
  max-height: 150px;
  max-width: 500px;
  min-height: 50px;
  overflow-y: scroll;
  padding: 5px 2px;
}

