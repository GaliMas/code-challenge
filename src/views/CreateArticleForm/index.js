import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactDom from 'react-dom';
import Modal from '../../components/Modal';
import Tag from '../../components/Tag';
import './index.css';

class CreateArticleForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      id: props.id || '',
      author: props.author || '',
      content: props.content || '',
      published: props.published || false,
      title: props.title || '',
      tags: props.tags || [],
    };
    this.handleChange = this.handleChange.bind(this);
    this.onKeyPress = this.onKeyPress.bind(this);
    this.tags = '';
  }

  onKeyPress(event) {
    if (event.which === 13) {
      this.setState({ tags: this.state.tags.concat(event.currentTarget.value) });
      ReactDom.findDOMNode(this.tags).value = '';
    }
  }

  handleChange(event) {
    const target = event.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const name = target.name;

    this.setState(prevState => ({ [name]: name !== 'tags' ? value : prevState.tags.concat(value) }));
  }

  render() {
    const { author, title, content, published } = this.props;
    return (
      <Modal onClose={this.props.onClose} title="New article">
        <div className="createArticleForm">
          <div className="formRow">
            <label className="labelForm">Author</label>
            <input defaultValue={author} name="author" type="text" onChange={this.handleChange} />
          </div>
          <div className="formRow">
            <label className="labelForm">Title</label>
            <input defaultValue={title} name="title" type="text" onChange={this.handleChange} />
          </div>
          <div className="formRow">
            <label className="labelForm">Content</label>
            <textarea defaultValue={content} name="content" type="text" onChange={this.handleChange} rows="5" />
          </div>
          <div className="formRow">
            <label className="published" id="published">Published</label>
            <input
              defaultChecked={published}
              className="published"
              name="published"
              type="checkbox"
              onChange={this.handleChange}
              htmlFor="published"
            />
          </div>
          <div className="formRow">
            <label className="tags">Tags</label>
            <input name="tags" type="text" onKeyPress={this.onKeyPress} ref={input => { this.tags = input; }} />
            {<div className="tagContainer">{this.state.tags.map((tag, i) => <Tag key={i}>{tag}</Tag>)}</div>}
          </div>
          <button type="button" className="btn cancel" onClick={this.props.onClose}>Cancelar</button>
          <button type="button" className="btn save" onClick={() => this.props.onSave(this.state)}>Guardar</button>
        </div>
      </Modal>
    );
  }
}

CreateArticleForm.propTypes = {
  author: PropTypes.string,
  content: PropTypes.string,
  id: PropTypes.string,
  onClose: PropTypes.func.isRequired,
  onSave: PropTypes.func.isRequired,
  published: PropTypes.bool,
  tags: PropTypes.arrayOf(PropTypes.string),
  title: PropTypes.string,
};

CreateArticleForm.defaultProps = {
  tags: [],
};

export default CreateArticleForm;
