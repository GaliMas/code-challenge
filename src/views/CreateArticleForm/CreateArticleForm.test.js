import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import CreateArticleForm from '.';

it('renders correctly', () => {
  const onClose = jest.fn();
  const onSave = jest.fn();
  const tree = renderer.create(
    <CreateArticleForm onClose={onClose} onSave={onSave} />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should trigger onClose', () => {
  const onClose = jest.fn();
  const onSave = jest.fn();
  const wrapper = mount(<CreateArticleForm onClose={onClose} onSave={onSave} />);
  wrapper.find('.cancel').simulate('click');
  expect(onClose.mock.calls.length).toBe(1);
});

it('should trigger onClose', () => {
  const onClose = jest.fn();
  const onSave = jest.fn();
  const wrapper = mount(<CreateArticleForm onClose={onClose} onSave={onSave} />);
  wrapper.find('.save').simulate('click');
  expect(onSave.mock.calls.length).toBe(1);
});
