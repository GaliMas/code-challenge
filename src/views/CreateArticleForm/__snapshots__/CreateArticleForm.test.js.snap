exports[`test renders correctly 1`] = `
<div
  className="modal undefined"
  id="myModal">
  <div
    className="modal-content">
    <header
      className="modalTitle">
      <span
        className="close"
        onClick={[Function]}>
        Ã—
      </span>
      <h2>
        New article
      </h2>
    </header>
    <div
      className="createArticleForm">
      <div
        className="formRow">
        <label
          className="labelForm">
          Author
        </label>
        <input
          defaultValue={undefined}
          name="author"
          onChange={[Function]}
          type="text" />
      </div>
      <div
        className="formRow">
        <label
          className="labelForm">
          Title
        </label>
        <input
          defaultValue={undefined}
          name="title"
          onChange={[Function]}
          type="text" />
      </div>
      <div
        className="formRow">
        <label
          className="labelForm">
          Content
        </label>
        <textarea
          defaultValue={undefined}
          name="content"
          onChange={[Function]}
          rows="5"
          type="text" />
      </div>
      <div
        className="formRow">
        <label
          className="published"
          id="published">
          Published
        </label>
        <input
          className="published"
          defaultChecked={undefined}
          htmlFor="published"
          name="published"
          onChange={[Function]}
          type="checkbox" />
      </div>
      <div
        className="formRow">
        <label
          className="tags">
          Tags
        </label>
        <input
          name="tags"
          onKeyPress={[Function]}
          type="text" />
        <div
          className="tagContainer" />
      </div>
      <button
        className="btn cancel"
        onClick={[Function]}
        type="button">
        Cancelar
      </button>
      <button
        className="btn save"
        onClick={[Function]}
        type="button">
        Guardar
      </button>
    </div>
  </div>
</div>
`;
